src/ivalue_manager.rs

```diff
@@ -63,11 +68,81 @@ fn update<F: FnMut(IValue) -> Result<Option<IValue>, Error>>(
                 if let Ok(x) = token.parse::<usize>() {
                     if is_last {
                         if x < arr.len() {
-                            let v = std::mem::replace(&mut arr[x], IValue::NULL);
-                            if let Some(res) = (func)(v)? {
-                                arr[x] = res;
-                            } else {
-                                arr.remove(x);
+                            let mut v = &mut arr.as_mut_slice()[x];
+                            match (func)(&mut v) {
+                                Ok(res) => {
+                                    if let Some(res) = res {
+                                        *v = res;
+                                    } else {
+                                        arr.remove(x);
+                                    }
+                                }
+                                Err(err) => return Err(err),
+                            }
+                        }
+                        return Ok(());
+                    }
+                    arr.get_mut(x)
+                } else {
+                    None
```

Comment:
Can this happened? if not asset.

Reply:
I believe that if something can not happened assert it, this way it will be easier to catch bugs.